text/0000-router-service.md

```diff
@@ -0,0 +1,440 @@
+
+Terminology:
+
+ - `EmberRouter` is the class that we already have today, defined in `ember-routing/system/router` and available publicly as `Ember.Router`
+ - `RouterService` is the new class I am proposing.
+
+`EmberRouter` has the following public API today:
+
+ - `map`
+ - `location`
+ - `rootURL`
+ - `willTransition`
+ - `didTransition`
+
+That API will be carried over verbatim to `RouterService`, and the publicly accessible `Ember.Router` class will *become* `RouterService`. In terms of implementation, I expect the existing `EmberRouter` class will continue to exist mostly unchanged. But public access to it will be moderated through `RouterService`.
+
+### New Methods: Initiating Transitions
+
+```js
+transitionTo(routeName, ...models, queryParams)
+replaceWith(routeName, ...models, queryParams)
+```
+
+These two have the same semantics as the existing methods on `Ember.Route`:
+
+### New Method: Checking For Active Route
+
+ - `isActive(routeName, ...models, queryParams)`
+
+The arguments have the same semantics as `transitionTo`, the return value is a boolean. This should provide the same logic that determines whether to put an active class on a `link-to`. Here's an example of how we can implement `is-active` as a helper, using this method:
+
+```js
+import Helper from 'ember-helper';
+import service from 'ember-service/inject';
+import observer from 'ember-metal/observer';
+
+export default Helper.extend({
+  router: service(),
+  compute([routeName, ...models], hash) {
+    let allModels;
+    if (hash.models) {
+      allModels = models.concat(hash.models);
+    } else {
+      allModels = models;
+    }
+    return this.get('router').isActive(routeName, ...models, hash.queryParams);
+  },
+  didTransition: observer('router.currentRoute', function() {
+    this.recompute();
+  });
+});
+```
+
+```hbs
+{{!- Example usage -}}
+<li class={{if (is-active "person.detail" model) 'chosen'}} >
+
+{{!- Example usage with generic routeName and list of models (avoids splat) -}}
+<a class={{if (is-active routeName models=models) 'chosen'}} >
+
+{{!- Note that the complexities of currentWhen can be avoided by composing instead. }}
+<a class={{if (or (is-active 'one') (is-active 'two')) 'active'}} href={{url-for 'two'}} >
+
+```
+
+### New Method: URL generation
+
+`urlFor(routeName, ...models, queryParams)`
+
+This takes the same arguments as `transitionTo`, but instead of initiating the transition it returns the resulting root-relative URL as a string (which will include the application's `rootUrl`).
+
+A `url-for` helper can be implemented almost identically to the `is-active` example above.
+
+### New Method: URL recognition
+
+`recognize(url)`
+
+Takes a string URL and returns a `RouteInfo` for the leafmost route represented by the URL. Returns `null` if the URL is not recognized. This method expects to receive the actual URL as seen by the browser _including_ the app's `rootURL`.
```

Comment:
In addition to returning `null` can `recognize` also call a hook on the `location` (e.g., `location.routeNotFound`)? 

One of the downsides of the current router is that it [throws an exception](https://github.com/tildeio/router.js/blob/b921fd3407064a77bd99f4f4eddc4377592ac93b/lib/router/transition-intent/url-transition-intent.js#L17-L22) when no route is found.

This makes it impossible to recover from an unrecognized route. In the case of FastBoot, it forces us to handle unrecognized routes at the server middleware layer, instead in the Ember application itself.

@tomdale or @danmcclain, Please add any other concerns that I might have forgot.